File: Example/HighFit/HighFit/HighFit/DataTableViewController.m
//
//  DataTableViewController.m
//  HighFit
//
//  License: www.highcharts.com/license
//  Copyright © 2023 Highsoft AS. All rights reserved.
//

#import "DataTableViewController.h"
#import "DataViewController.h"
#import <Highcharts/Highcharts.h>
#import "DashboardViewController.h"
#import "OptionsProvider.h"

@interface DataTableViewController ()
@property (strong, nonatomic) NSString *chartType;
@property (strong, nonatomic) NSDictionary *data;
@property (strong, nonatomic) UIView *chartViewBase;
@property (strong, nonatomic) HIChartView *chartView;

@property (strong, nonatomic) UISwitch *switchView;
@end

@implementation DataTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
    
    self.chartType = self.configuration[@"chartType"];
    
    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStyleGrouped];
    
    NSData *data = [NSData dataWithContentsOfFile:[[NSBundle mainBundle] pathForResource:self.configuration[@"source"] ofType:@"json"]];
    NSError *error = nil;
    self.data = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&error];
    
    self.title = self.configuration[@"title"];
    
    self.tableView.tableHeaderView = self.toolbar;
    
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    
    [self.segment addTarget:self action:@selector(actionSegment:) forControlEvents:UIControlEventValueChanged];
    
    self.switchView = [[UISwitch alloc] init];
    [self.switchView addTarget:self action:@selector(actionSwitch:) forControlEvents:UIControlEventValueChanged];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    self.switchView.on = [self isSwitchOn];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 3;
}

- (UIView*)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
{
    if (!self.chartViewBase) {
        
        self.chartViewBase = [[UIView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, self.view.frame.size.width, 260.0f)];
        self.chartViewBase.backgroundColor = [UIColor whiteColor];
        NSMutableDictionary *tmpOptions = [NSMutableDictionary dictionaryWithDictionary:self.configuration];
        tmpOptions[@"exporting"] = @YES;
        self.chartView = [[HIChartView alloc] initWithFrame:CGRectMake(5.0f, 5.0f, self.view.frame.size.width-20, 240.0f)];
        int sum = 0;
        for (NSNumber *number in self.data[@"day"]) {
            sum += number.integerValue;
        }
        
        tmpOptions[@"subtitle"] = [NSString stringWithFormat:@"%d %@", sum, tmpOptions[@"unit"]];

        self.chartView.options = [OptionsProvider provideOptionsForChartType:tmpOptions series:self.data[@"day"] type:@"day"];
        self.chartView.viewController = self;
        
        [self.chartViewBase addSubview:self.chartView];
    }
    
    return self.chartViewBase;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    return 260.0f;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"HighFitCellData"];
    
    // Configure the cell...
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"HighFitCellData"];
    }
    
    cell.accessoryView = nil;
    cell.textLabel.text = nil;
    cell.accessoryType = UITableViewCellAccessoryNone;
    
    if (indexPath.row == 0) {
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        cell.textLabel.text = @"Show on Dashboard.";
        
        cell.accessoryView = self.switchView;
    }
    
    if (indexPath.row == 1) {
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        cell.textLabel.text = @"Show all Data";
    }
    
    if (indexPath.row == 2) {
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        cell.textLabel.text = @"Unit";
        
        UILabel *label = [[UILabel alloc] init];
        [label setText:self.configuration[@"unit"]];
        [label setFont:[UIFont systemFontOfSize:16.0f]];
        [label setTextColor:[UIColor lightGrayColor]];
        [label sizeToFit];
        
        cell.accessoryView = label;
    }
    
    return cell;
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

#pragma mark - Table view delegate

// In a xib-based application, navigation from a table can be handled in -tableView:didSelectRowAtIndexPath:
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    // Navigation logic may go here, for example:
    // Create the next view controller.
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    if (indexPath.row != 1) {
        return;
    }
    
    DataViewController *detailViewController = [[DataViewController alloc] initWithNibName:@"DataViewController" bundle:nil];
    
    // Pass the selected object to the new view controller.
    detailViewController.unit = self.configuration[@"unit"];
    detailViewController.data = self.data[@"all"];
    
    // Push the view controller.
    [self.navigationController pushViewController:detailViewController animated:YES];
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

#pragma mark - Additinal methods

- (void)actionSwitch:(UISwitch*)actionSwitch
{
    // Show or hide chart on dashboard.
    if (actionSwitch.isOn) {
        [[DashboardViewController sharedDashboard] dataSourceAdd: self.configuration];
    } else {
        [[DashboardViewController sharedDashboard] dataSourceRem: self.configuration];
    }
}

- (IBAction)actionSegment:(UISegmentedControl*)sender
{
    NSString *dataName = @"day";
    
    switch (sender.selectedSegmentIndex) {
        case 0:
            dataName = @"day";
            break;
        case 1:
            dataName = @"week";
            break;
        case 2:
            dataName = @"month";
            break;
        case 3:
            dataName = @"year";
            break;
        default:
            break;
    }
    
    NSMutableDictionary *tmpOptions = [NSMutableDictionary dictionaryWithDictionary:self.configuration];
    tmpOptions[@"exporting"] = @YES;
    
    int sum = 0;
    for (NSNumber *number in self.data[dataName]) {
        sum += number.integerValue;
    }
    
    tmpOptions[@"subtitle"] = [NSString stringWithFormat:@"%d %@", sum, tmpOptions[@"unit"]];
    
    self.chartView.options = [OptionsProvider provideOptionsForChartType:tmpOptions series:self.data[dataName] type:dataName];
}

- (BOOL)isSwitchOn
{
    NSArray *sources = [[NSUserDefaults standardUserDefaults] valueForKey:@"sources"];
    for (NSDictionary *item in sources) {
        if ([[item valueForKey:@"source"] isEqualToString:self.configuration[@"source"]]) {
            return YES;
        }
    }
    return NO;
}

@end

File: Example/HighFit/HighFit Swift/HighFit/DataTableViewController.swift
//
//  DataTableViewController.swift
//  HighFit
//
//  License: www.highcharts.com/license
//  Copyright © 2020 Highsoft AS. All rights reserved.
//

import UIKit
import Highcharts

class DataTableViewController: UITableViewController, HIChartViewDelegate {
    
    @IBOutlet var toolbar: UIToolbar!
    @IBOutlet weak var segment: UISegmentedControl!
    
    var configuration: [String: Any]!
    
    var chartType: String!
    var data: [String : Any]!
    var chartViewBase: UIView!
    var chartView: HIChartView!
    
    var switchView: UISwitch!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false
        
        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem()
        
        self.chartType = (self.configuration["chartType"] as! String)
        
        self.tableView = UITableView(frame: self.view.bounds, style: .grouped)
        
        do {
            if let sourceName = self.configuration["source"] as? String,
                let sourcePath = Bundle.main.path(forResource: sourceName, ofType: "json"),
                let sourceData = try? Data(contentsOf: URL(fileURLWithPath: sourcePath)),
                let sourceJson = try JSONSerialization.jsonObject(with: sourceData) as? [String: Any] {
                self.data = sourceJson
            }
        } catch {
            print("Error deserializing JSON: \(error)")
        }
        
        self.title = self.configuration["title"] as? String
        
        self.tableView.tableHeaderView = self.toolbar
        
        self.tableView.tableFooterView = UIView(frame: CGRect.zero)
        
        self.segment.addTarget(self, action: #selector(self.actionSegment), for: .valueChanged)
        
        self.switchView = UISwitch()
        self.switchView.addTarget(self, action: #selector(self.actionSwitch), for: .valueChanged)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        self.switchView.isOn = self.isSwitchOn()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Table view data source
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 3
    }
    
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        if self.chartViewBase == nil {
            self.chartViewBase = UIView(frame: CGRect(x: 0.0, y: 0.0, width: self.view.frame.size.width, height: 260.0))
            self.chartViewBase.backgroundColor = UIColor.white
            
            var tmpOptions = self.configuration!
            tmpOptions["exporting"] = true
            
            self.chartView = HIChartView(frame: CGRect(x: 5.0, y: 5.0, width: self.view.frame.size.width - 20, height: 240.0))
            self.chartView.delegate = self
            
            let series = self.data["day"] as! [Int]
            var sum: Int = 0
            for number in series {
                sum += number
            }
            
            tmpOptions["subtitle"] = "\(sum) \(tmpOptions["unit"]!)"
            
            self.chartView.options = OptionsProvider.provideOptions(forChartType: tmpOptions, series: series, type: "day")
            self.chartView.viewController = self
            
            self.chartViewBase.addSubview(self.chartView!)
        }
        return self.chartViewBase
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 260.0
    }
    
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: "HighFitCellData")
        
        // Configure the cell...
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: "HighFitCellData")
        }
        
        cell!.accessoryView = nil
        cell!.textLabel?.text = nil
        cell!.accessoryType = .none
        
        if indexPath.row == 0 {
            cell!.selectionStyle = .none
            cell!.textLabel?.text = "Show on Dashboard."
            
            cell!.accessoryView = self.switchView
        }
        
        if indexPath.row == 1 {
            cell!.accessoryType = .disclosureIndicator
            cell!.textLabel?.text = "Show all Data"
        }
        
        if indexPath.row == 2 {
            cell!.selectionStyle = .none
            cell!.textLabel?.text = "Unit"
            
            let label = UILabel()
            label.text = self.configuration["unit"] as? String
            label.font = UIFont.systemFont(ofSize: 16.0)
            label.textColor = UIColor.lightGray
            label.sizeToFit()
            cell!.accessoryView = label
        }
        
        return cell!
    }
    
    /*
     // Override to support conditional editing of the table view.
     override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
     // Return false if you do not want the specified item to be editable.
     return true
     }
     */
    
    /*
     // Override to support editing the table view.
     override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
     if editingStyle == .delete {
     // Delete the row from the data source
     tableView.deleteRows(at: [indexPath], with: .fade)
     } else if editingStyle == .insert {
     // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
     }
     }
     */
    
    /*
     // Override to support rearranging the table view.
     override func tableView(_ tableView: UITableView, moveRowAt fromIndexPath: IndexPath, to: IndexPath) {
     
     }
     */
    
    /*
     // Override to support conditional rearranging of the table view.
     override func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
     // Return false if you do not want the item to be re-orderable.
     return true
     }
     */
    
    // MARK: - Table view delegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        if indexPath.row != 1 {
            return
        }
        
        let detailViewController = DataViewController(nibName: "DataViewController", bundle: nil)
        
        // Pass the selected object to the new view controller.
        detailViewController.unit = self.configuration["unit"] as? String
        detailViewController.data = (self.data["all"] as! [Any])
        
        // Push the view controller.
        self.navigationController?.pushViewController(detailViewController, animated: true)
    }
    
    /*
     // MARK: - Navigation
     
     // In a storyboard-based application, you will often want to do a little preparation before navigation
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
     // Get the new view controller using segue.destinationViewController.
     // Pass the selected object to the new view controller.
     }
     */
    
    // MARK: - Additional methods
    
    @objc func actionSwitch(_ actionSwitch: UISwitch) {
        // Show or hide chart on dashboard.
        if actionSwitch.isOn {
            DashboardViewController.sharedDashboard().dataSourceAdd(self.configuration)
        }
        else {
            DashboardViewController.sharedDashboard().dataSourceRem(self.configuration)
        }
    }
    
    @IBAction func actionSegment(_ sender: UISegmentedControl) {
        var dataName: String = "day"
        
        switch sender.selectedSegmentIndex {
        case 0:
            dataName = "day"
        case 1:
            dataName = "week"
        case 2:
            dataName = "month"
        case 3:
            dataName = "year"
        default:
            break
        }
        
        
        var tmpOptions = self.configuration!
        tmpOptions["exporting"] = true
        
        let series = self.data[dataName] as! [Int]
        var sum: Int = 0
        for number in series {
            sum += number
        }
        
        tmpOptions["subtitle"] = "\(sum) \(tmpOptions["unit"]!)"
        
        self.chartView.options = OptionsProvider.provideOptions(forChartType: tmpOptions, series: series, type: dataName)
    }
    
    private func isSwitchOn() -> Bool {
        let sources = UserDefaults.standard.value(forKey: "sources") as! [[String:Any]]
        for item in sources {
            if item["source"] as! String == self.configuration["source"] as! String {
                return true
            }
        }
        return false
    }
    
    //MARK: - HIChartViewDelegate
    
    func chartViewDidLoad(_ chart: HIChartView!) {
        print("Did load chart \(chart!)")
    }
    
}
